### PR TITLE
8311542: Consolidate the native stack printing code

### PR DESCRIPTION
**Pull request by dholmes-ora**
We now print native stacks in a number of contexts, not just VMError reporting, and we have to try `os::platform_print_native_stack` else fall back to `VMError::print_native stack`. 

The refactoring adds a new `NativeStackPrinter` (NSP) helper class which can be constructed with some of the context information for the printing that will follow. This avoids the need for the print functions to have a large number of parameters. We have to expose both the top-level printing functionality and the "lower-level" frame-based stack walk as the error reporter needs access to that directly. To maintain the exact same format of output the NSP has to be aware of some error reporter usage requirements.

I also had to expose a direct `frame` taking print function for the Debug.cpp `pns` case.

Testing:
 - tiers 1 - 4

Some frame management code was also moved from `VMError` to the `frame` class.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311542](https://bugs.openjdk.org/browse/JDK-8311542): Consolidate the native stack printing code (**Enhancement** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22472/head:pull/22472` \
`$ git checkout pull/22472`

Update a local copy of the PR: \
`$ git checkout pull/22472` \
`$ git pull https://git.openjdk.org/jdk.git pull/22472/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22472`

View PR using the GUI difftool: \
`$ git pr show -t 22472`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22472.diff">https://git.openjdk.org/jdk/pull/22472.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22472#issuecomment-2511361230)
</details>

### CODE REVIEW

File: src/hotspot/share/runtime/frame.cpp
Comment (kimbarrett) [inline src/hotspot/share/runtime/frame.cpp:1566]: `frame` is a non-trivial type to copy, so it seems like it would be better to pass it by const-ref rather
than by value.  That might trip over lots of const-correctness issues when accessing it though.  If
pulling that thread looks like an excessive tangle then clean this up as a followup.

File: src/hotspot/share/utilities/nativeStackPrinter.cpp
Comment (kimbarrett) [inline src/hotspot/share/utilities/nativeStackPrinter.cpp:44]: Why is `frame` passed by non-const reference?  This doesn't seem like a function that should be modifying
it?  Though changing this might also run afoul of const-correctness issues when accessing the frame, so
feel free to defer to a followup if that's a problem.

**Review by kimbarrett — APPROVED** [review]
Looks good, except for some const issues.  I'm approving as is if you decide not to
address those in this PR.

File: src/hotspot/share/runtime/frame.cpp
Comment (dholmes-ora) [reply src/hotspot/share/runtime/frame.cpp:1566]: It seems we often pass frames around rather than pointers or references to them, but I don't know why. So I would not want to be changing any existing signatures in case there is some nuance I am not aware of.

File: src/hotspot/share/utilities/nativeStackPrinter.cpp
Comment (dholmes-ora) [reply src/hotspot/share/utilities/nativeStackPrinter.cpp:44]: I passed by reference here just to avoid adding in an additional level of copying in the new code as I was unsure if an additional copy might have an adverse impact. I never considered applying const and I also suspect it will have a significant flow-on effect if I try.

**Comment by dholmes-ora** [comment]
Thanks for the review @kimbarrett !

File: src/hotspot/share/utilities/nativeStackPrinter.cpp
Comment (stefank) [reply src/hotspot/share/utilities/nativeStackPrinter.cpp:44]: Note that the contents in fr is overwritten in this function. All callers of this function must be careful because of that. I vote for fixing it before pushing this.

**Review by stefank — CHANGES_REQUESTED** [review]

**Review by TheShermanTanker — APPROVED** [review]
*Cries in merge conflicts galore

File: src/hotspot/share/utilities/nativeStackPrinter.cpp
Comment (dholmes-ora) [reply src/hotspot/share/utilities/nativeStackPrinter.cpp:44]: @stefank - yes you are right. I think I introduced the reference in an earlier version of the code to avoid introducing an intermediary frame object, but as it stands the frame is passed in directly whereas the original code made a copy. I will fix it. Thanks.

File: src/hotspot/share/runtime/frame.cpp
Comment (dholmes-ora) [reply src/hotspot/share/runtime/frame.cpp:1566]: @fbredber  thanks for looking at this. I'm not trying to change anything here, all I did was move the code from VMError into frame. Terminology gets complex with frames and we already have the notion of "sender" rather than caller in this code.

**Comment by TheShermanTanker** [comment]
> > *Cries in merge conflicts galore
> 
> ??? Sorry that was too cryptic for me.

Oh, it was just a joke about how it would conflict with what I do downstream to use print\_native\_stack instead of the Windows platform\_print\_native\_stack, although I just checked my port and it seems like this change doesn't actually conflict with it :)

File: src/hotspot/share/utilities/debug.cpp
Comment (TheShermanTanker) [inline src/hotspot/share/utilities/debug.cpp:653]: Should this be aligned with tty or is the current formatting preferred? Alternatively it seems like this could all fit on one line, but it's up to you

File: src/hotspot/share/runtime/frame.hpp
Comment (TheShermanTanker) [inline src/hotspot/share/runtime/frame.hpp:443]: A thought: Does it make more sense to have next\_frame implemented in vmError or nativeStackPrinter and made available through their respective headers rather than as a member of the frame class? Just thinking out loud, I'm fine with it as is

File: src/hotspot/share/runtime/frame.cpp
Comment (TheShermanTanker) [inline src/hotspot/share/runtime/frame.cpp:1564]: I got a bit of a chuckle looking at this, the triple equals reminds me of the wonky JavaScript equality checking. Not a review, I just found it funny :)

**Review by TheShermanTanker — APPROVED** [review]
Looks good

File: src/hotspot/share/utilities/debug.cpp
Comment (dholmes-ora) [reply src/hotspot/share/utilities/debug.cpp]: Fixed - thanks.

File: src/hotspot/share/runtime/frame.hpp
Comment (dholmes-ora) [reply src/hotspot/share/runtime/frame.hpp:443]: I figured the only class that should know that much about how to walk frames should the frame class. In addition it would look for either of the other classes to export that kind of API for other classes to use.

Thanks for re-reviewing. Of course it needs another review now I fixed the indentation issue.

File: src/hotspot/share/runtime/frame.cpp
Comment (dholmes-ora) [reply src/hotspot/share/runtime/frame.cpp:1564]: I never noticed the triple :)

File: src/hotspot/share/runtime/frame.hpp
Comment (TheShermanTanker) [reply src/hotspot/share/runtime/frame.hpp:443]: Ah, I was talking about having it as a top level extern method in either nativeStackPrinter or vmError, and declared in their headers for other code to use, not as a member of any of their classes, but I guess it doesn't matter. Thanks for your patience

I don't think my review would've mattered anyway even if you hadn't changed anything since I'm not a Reviewer, but for the sake of it I'll re\-approve :)

**Review by TheShermanTanker — APPROVED** [review]

**Comment by TheShermanTanker** [comment]
Thanks for the fixes!

**Comment by dholmes-ora** [comment]
Thanks @kimbarrett !

/integrate

**Pull request closed without merging**
